### ISSUE TITLE
Children rendered in the wrong order unless both of them are always rendered

### ISSUE THREAD

**Issue by fabiospampinato** [preactjs/preact#3037]
### Reproduction

I was unable to make a repro unfortunately.

### Steps to reproduce

I'll start with a video to make this clearer, basically I'm toggling a component ("Viewbar") in an app and when that component becomes rendered after it has been non-rendered (return `null`) it becomes rendered in the wrong position:

https://user-images.githubusercontent.com/1812093/108720605-efeec680-7518-11eb-9887-9df4e3ebbf39.mov

- This bug is not present when using React.
- I can't reproduce it just by toggling one of two child components in a codepen.
	- Not even when managing the state for that with the state management [library](https://github.com/fabiospampinato/store) that I use in the app.

Any ideas on how to debug this? I'm on preact/compat latest.

### Expected Behavior

The order of components should be stable even after any of them returned `null`.

### Actual Behavior

The order is not stable.

**Comment by fabiospampinato** [comment]
I've kind of debugged this, basically I have a component, let's call it "Layout", which can accept some children components, the thing that Layout does is making the boundaries between those children draggable so that the children themselves can be resized, it's basically a component that manages resizable columns.

Now the issue is that in order for Layout to make those boundaries draggable it adds the necessary nodes to the DOM that handle the dragging, having those extra nodes added just throws off Preact, but it causes no problems with React.

This way of implementing resizable columns may sound a little strange maybe, but if you can't introduce those draggable nodes automatically then maybe you'll have to do it manually, but you can't really do it manually without breaking encapsulation because you don't know whether the children will actually render something or not, so maybe the only option left is to not use elements at all, and maybe you'll have to rely on pseudo elements, but that has drawbacks too, so IMHO the best approach is to add those nodes automatically.

Bottom line is Preact is buggy in this regard compared to React, if this sounds like something you guys want to fix I can try to make a repro.

**Comment by fabiospampinato** [comment]
Interestingly when the column is first hidden there are no such draggable nodes inserted in the DOM (no boundaries), so when it's asked to render again there aren't even any of those draggable nodes present yet, yet for some reason Preact still puts the now visible columns in the wrong position.

**Comment by fabiospampinato** [comment]
I tried making a clean repro but for some reason I still can't reproduce the issue with it, [here](https://codepen.io/FabioSpampinato/pen/vYyeRdb)'s the code for what is worth.

I tried extracting a repro from my codebase but things are too tangled with each other for me to be able to do that in any reasonable amount of time.

So I don't know, maybe I can connect with somebody on TeamViewer or something, or I can try making a repro again some other time. For now unfortunately the only useful bits of information that I have are that the bug disappears by wrapping children in a fragment, and [here] there's the underlying code for that Layout component, which is probably related to the bug in question because just using a dummy wrapper component makes the issue disappear too.

**Comment by jamesscottbrown** [comment]
I also hit this bug (or, at least, a similar bug), and have a way to reproduce it:

* create a new project with ` npx preact-cli create simple dom-bug-reprex`
* replace `index.js` with 

```js
import {useState} from 'preact/hooks'


const Item = ({array, item, setData}) => {

    const remove = () => setData(array.filter(d => d !== item));

    return <div>
		<span>{item} <button onclick={remove}>remove</button> </span>
    </div>
}


export default function App() {

	const [a, setA] = useState([1,2]);
	const [b, setB] = useState([3,4]);
	const [c, setC] = useState([5,6]);

	return <>
            <h1>Welcome!</h1>

            {
                a.length > 0 && <>
                    <h2>First Section</h2>
                    {a.map(g => <Item item={g} array={a} setData={setA} />)}
                </>
            }

            {
                b.length > 0 && <>
                    <h2>Second Section</h2>
                    {b.map(g => <Item item={g} array={b} setData={setB} />)}
                </>
            }

            {
                c.length > 0 && <>
                    <h2>Third Section</h2>
                    {c.map(g => <Item item={g} array={c} setData={setC} />)}
                </>
            }
        </>
}
```

Pressing the "Remove" button next to any of 1/2/3/4 will mess up the order of the sections.

As @fabiospampinato  observed, this can be worked-around by using `<>` and `</>`  to give:


```js

export default function App() {

    const [a, setA] = useState([1, 2]);
    const [b, setB] = useState([3, 4]);
    const [c, setC] = useState([5, 6]);

    return <>
        <h1>Welcome!</h1>

        <>
            {
                a.length > 0 && <>
                    <h2>First Section</h2>
                    {a.map(g => <Item item={g} array={a} setData={setA}/>)}
                </>
            }

            {
                b.length > 0 && <>
                    <h2>Second Section</h2>
                    {b.map(g => <Item item={g} array={b} setData={setB}/>)}
                </>
            }

            {
                c.length > 0 && <>
                    <h2>Third Section</h2>
                    {c.map(g => <Item item={g} array={c} setData={setC}/>)}
                </>
            }
        </>
    </>
}

```

**Comment by jamesscottbrown** [comment]
Setting a `key` for each `item` doesn't solve the problem, but does have an effect - it fixes the problem for the delete button next to 1, but not the delete button next to 2.

**Comment by jamesscottbrown** [comment]
This might be the same issue as https://github.com/preactjs/preact/issues/2987 ?

**Comment by fabiospampinato** [comment]
Closing as `latest` fixed it.
